src/components/Panel.ts

```diff
@@ -74,6 +80,170 @@ class Panel {
     this.state.cachedBbox = null;
   }
 
+  public getProgress() {
+    const viewport = this.viewport;
+    const options = viewport.options;
+    const panelCount = viewport.panelManager.getPanelCount();
+    const scrollAreaSize = viewport.getScrollAreaSize();
+
+    const relativeIndex = (options.circular ? Math.floor(this.getPosition() / scrollAreaSize) * panelCount : 0) + this.getIndex();
+    const progress = relativeIndex - viewport.getCurrentProgress();
+
+    return progress;
+  }
+
+  public getOutsetProgress() {
+    const viewport = this.viewport;
+    const outsetRange = [
+      -this.getSize(),
+      viewport.getRelativeHangerPosition() - this.getRelativeAnchorPosition(),
+      viewport.getSize(),
+    ];
+    const relativePanelPosition = this.getPosition() - viewport.getCameraPosition();
+    const outsetProgress = getProgress(relativePanelPosition, outsetRange);
+
+    return outsetProgress;
+  }
+
+  public getVisibleRatio() {
+    const viewport = this.viewport;
+    const panelSize = this.getSize();
+    const relativePanelPosition = this.getPosition() - viewport.getCameraPosition();
+    const rightRelativePanelPosition = relativePanelPosition + panelSize;
+
+    const visibleSize = Math.min(viewport.getSize(), rightRelativePanelPosition) - Math.max(relativePanelPosition, 0);
+    const visibleRatio = visibleSize >= 0
+      ? visibleSize / panelSize
+      : 0;
+
+    return visibleRatio;
+  }
+
+  public focus(duration?: number): void {
+    const viewport = this.viewport;
+    const currentPanel = viewport.getCurrentPanel();
+    const hangerPosition = viewport.getHangerPosition();
+    const anchorPosition = this.getAnchorPosition();
+    if (hangerPosition === anchorPosition || !currentPanel) {
+      return;
+    }
+
+    const currentPosition = currentPanel.getPosition();
+    const eventType = currentPosition === this.getPosition()
+      ? ""
+      : EVENTS.CHANGE;
+
+    console.log(duration);
```

Comment:
log remained.

Reply:
That file's outdated, and fixed on later commit.